### PR TITLE
feat: add automatic in-place migrations for 02-client-refactor (v7)

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

ref: #1863 

### Commit Message / Changelog Entry

```bash
feat: add automatic in-place store migrations to remove the localhost client and migrate existing solo machine definitions.
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/24-host/parse.go [hunk 1]
```diff
@@ -32,6 +32,29 @@ func ParseIdentifier(identifier, prefix string) (uint64, error) {
 	return sequence, nil
 }
 
+// ParseClientStatePath returns the client ID from a client state path. It returns
+// an error if the provided path is invalid.
+func ParseClientStatePath(path string) (string, error) {
```
Comment (colin-axner) [inline modules/core/24-host/parse.go:37]: going to pull out into separate pr but needed for tests to run

File: modules/core/02-client/migrations/v7/solomachine.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
+package v7
```
Comment (colin-axner) [inline modules/core/02-client/migrations/v7/solomachine.go:1]: copied from `legacy/v100/solomachine.go`

Needed in order to unmarshal the existing types which no longer register solomachine v2 on the codec. The SDK codec will prematurely try to lookup type registration against the interface (even if the cdc isn't supposed to unpack the interface).

File: modules/core/02-client/migrations/v7/store.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		keySplit := strings.Split(string(iterator.Key()), "/")
+		// key is in the format "consensusStates/<height>"
+		if len(keySplit) != 2 || keySplit[0] != string(host.KeyConsensusStatePrefix) {
+			continue
+		}
+
+		// collect consensus states to be pruned
+		heights = append(heights, clienttypes.MustParseHeight(keySplit[1]))
```
Comment (colin-axner) [inline modules/core/02-client/migrations/v7/store.go:196]: existing code from `legacy/v100/store.go`, it can be cleaned up in a different pr (matches iteration in keeper functions)

File: modules/core/02-client/migrations/v7/solomachine.go [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+// The same is done for the ConsensusState.
+
+// Interface implementation checks.
+var (
+	_, _ codectypes.UnpackInterfacesMessage = &ClientState{}, &ConsensusState{}
+	_    exported.ClientState               = (*ClientState)(nil)
+	_    exported.ConsensusState            = &ConsensusState{}
+)
+
+func RegisterInterfaces(registry codectypes.InterfaceRegistry) {
```
Comment (damiannolan) [inline modules/core/02-client/migrations/v7/solomachine.go:30]: nit, may as well add a godoc

```suggestion

// RegisterInterfaces registers the solomachine v2 ClientState and ConsensusState types in the interface registry.
func RegisterInterfaces(registry codectypes.InterfaceRegistry) {
```

File: modules/core/02-client/migrations/v7/store.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+		any := &codectypes.Any{}
+		if err := cdc.Unmarshal(bz, any); err != nil {
+			return sdkerrors.Wrap(err, "failed to unmarshal client state bytes into solo machine client state")
+		}
+
+		clientState := &ClientState{}
+		if err := cdc.Unmarshal(any.Value, clientState); err != nil {
+			return sdkerrors.Wrap(err, "failed to unmarshal client state bytes into solo machine client state")
+		}
+
```
Comment (damiannolan) [inline modules/core/02-client/migrations/v7/store.go:76]: I think we normally favour unmarshalling into types like this in other parts of the codebase.
But I see the `ClientState` is a passed as a pointer type to `migrateSolomachine` also.

Happy to leave as is seeing as this is migration code anyways.

```suggestion
		var any codectypes.Any
		if err := cdc.Unmarshal(bz, &any); err != nil {
			return sdkerrors.Wrap(err, "failed to unmarshal client state bytes into solo machine client state")
		}

		var clientState ClientState
		if err := cdc.Unmarshal(any.Value, &clientState); err != nil {
			return sdkerrors.Wrap(err, "failed to unmarshal client state bytes into solo machine client state")
		}

```

File: modules/core/02-client/migrations/v7/store.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+		path := string(iterator.Key())
+		if !strings.Contains(path, host.KeyClientState) {
+			// skip non client state keys
+			continue
+		}
+
+		clientID := host.MustParseClientStatePath(path)
+		clients = append(clients, clientID)
+
+		// optimization: exist after a single tendermint client iteration
```
Comment (damiannolan) [inline modules/core/02-client/migrations/v7/store.go:168]:
```suggestion
		// optimization: exit after a single tendermint client iteration
```

File: modules/core/02-client/migrations/v7/store.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+			return clients, nil
+		}
+	}
+
+	return clients, nil
+}
+
+// pruneClientConsensusStates removes all client consensus states from the associated
+// client store.
+func pruneClientConsensusStates(clientStore sdk.KVStore) {
```
Comment (damiannolan) [inline modules/core/02-client/migrations/v7/store.go:179]: Should we consider renaming this function? Pruning suggests that something is kept/leftover, whereas this is effectively removing the entire localhost consensus state history right?

File: modules/core/02-client/migrations/v7/store_test.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+	err := v7.MigrateStore(suite.chainA.GetContext(), suite.chainA.GetSimApp().GetKey(host.StoreKey), cdc)
+	suite.Require().Error(err)
+
+	ibctm.RegisterInterfaces(registry)
+	err = v7.MigrateStore(suite.chainA.GetContext(), suite.chainA.GetSimApp().GetKey(host.StoreKey), cdc)
+	suite.Require().NoError(err)
+}
+
+// create multiple solo machine clients, tendermint and localhost clients
+// ensure that solo machine clients are migrated and their cosnensus states are removed
```
Comment (damiannolan) [inline modules/core/02-client/migrations/v7/store_test.go:61]:
```suggestion
// ensure that solo machine clients are migrated and their consensus states are removed
```

File: modules/core/02-client/migrations/v7/store_test.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+	err := v7.MigrateStore(suite.chainA.GetContext(), suite.chainA.GetSimApp().GetKey(host.StoreKey), suite.chainA.App.AppCodec())
+	suite.Require().NoError(err)
+
+	suite.assertSolomachineClients(solomachines)
+	suite.assertLocalhostClients()
+}
+
+func (suite *MigrationsV7TestSuite) createSolomachineClients(solomachines []*ibctesting.Solomachine) {
+	// manually generate old proto buf definitions and set in store
```
Comment (damiannolan) [inline modules/core/02-client/migrations/v7/store_test.go:90]:
```suggestion
	// manually generate old protobuf definitions and set in store
```

**Review by damiannolan — APPROVED** [review]
Excellent work, really clean! Left mostly nits.
LGTM 🚀

File: modules/core/02-client/keeper/migrations.go [hunk 10]
```diff
@@ -25,3 +26,13 @@ func NewMigrator(keeper Keeper) Migrator {
...
 	return v100.MigrateStore(ctx, m.keeper.storeKey, m.keeper.cdc)
 }
+
+// Migrate2to3 migrates from version 2 to 3.
+// This migration
+// - migrates solo machine client states from v2 to v3 protobuf definition
+// - prunes solo machine consensus states
+// - removes the localhost client
+// - asserts that existing tendermint clients are properly registered on the chain codec
+func (m Migrator) Migrate2to3(ctx sdk.Context) error {
```
Comment (chatton) [inline modules/core/02-client/keeper/migrations.go:36]: I have a question about the `Migrator` struct, do we need to this to be a struct (maybe implementing some interface?)

It looks like we only use 2 fields from the keeper and don't actually use any of the other keeper functionality. (cdc/storeKey)

I'm happy to stick with the pattern of adding functions, just curious if we need this struct!

File: modules/core/02-client/migrations/v7/store.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+	if err := handleTendermintMigration(ctx, store, cdc); err != nil {
+		return err
+	}
+
+	if err := handleLocalhostMigration(ctx, store, cdc); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (chatton) [inline modules/core/02-client/migrations/v7/store.go:48]: very comprehensive doc string and super clean/simple implementation, I love it! 🥇

File: modules/core/02-client/migrations/v7/store.go [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+	}
+
+	clientID := clients[0]
+
+	clientPrefix := []byte(fmt.Sprintf("%s/%s/", host.KeyClientStorePrefix, clientID))
+	clientStore := prefix.NewStore(store, clientPrefix)
+
+	bz := clientStore.Get(host.ClientStateKey())
+	if bz == nil {
+		return clienttypes.ErrClientNotFound
```
Comment (chatton) [inline modules/core/02-client/migrations/v7/store.go:112]: should we wrap this error in the same fashion as the other error cases? We could include information about the client state key in the error message.

File: modules/core/02-client/migrations/v7/store_test.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	suite.createSolomachineClients(solomachines)
+	suite.createLocalhostClients()
+
+	err := v7.MigrateStore(suite.chainA.GetContext(), suite.chainA.GetSimApp().GetKey(host.StoreKey), suite.chainA.App.AppCodec())
+	suite.Require().NoError(err)
+
+	suite.assertSolomachineClients(solomachines)
+	suite.assertLocalhostClients()
```
Comment (chatton) [inline modules/core/02-client/migrations/v7/store_test.go:86]: [nit] I think it would be nice to have what is being asserted clear in the testcase, i.e. `suite.assertNumLocalhostClients(0)`  or `suite.assertNoLocalhostClients()` rather than digging into the function to see what it is doing.

**Review by chatton — APPROVED** [review]
Great job, the code is super clean, I just had a few small nits/questions!

File: modules/core/02-client/keeper/migrations.go (hunk 10 above)
Comment (colin-axner) [reply modules/core/02-client/keeper/migrations.go:36]: This is a good question. This was the approach documented by the [SDK](https://docs.cosmos.network/main/building-modules/upgrade)

> Since these migrations are functions that need access to a Keeper's store, use a wrapper around the keepers called Migrator as shown in this example:

I think it makes sense for the Keeper not to expose `Migrate2to3`, but I'm totally open to a different design

File: modules/core/02-client/migrations/v7/store_test.go (hunk 13 above)
Comment (colin-axner) [reply modules/core/02-client/migrations/v7/store_test.go]: agreed! Renamed to `assertNoLocalhostClients`

File: modules/core/02-client/migrations/v7/store.go (hunk 5 above)
Comment (colin-axner) [reply modules/core/02-client/migrations/v7/store.go]: applied!
